// simply-xp/src/rank.js
const levels = require('../src/models/level.js')
const { join } = require('path')


/**
 * @param {Discord.Message} message
 * @param {string} userID
 * @param {string} guildID
 * @param {import('../index').rankOptions} options
 */

async function rank(message, userID, guildID, options = []) {
  if (!userID) throw new Error('[XP] User ID was not provided.')

  if (!guildID) throw new Error('[XP] Guild ID was not provided.')

  const user = await levels.findOne({
    user: userID,
    guild: guildID
  })
  if (!user) throw new Error('[XP] NO_DATA | User has no XP data.')

  const leaderboard = await levels
    .find({
      guild: guildID
    })
    .sort([['xp', 'descending']])
    .exec()

  user.position = leaderboard.findIndex((i) => i.user === userID) + 1

  let targetxp = user.level + 1

  let target = targetxp * targetxp * 100

  return rankCard(message, {
    level: user.level,
    currentXP: user.xp,
    neededXP: target,
    rank: user.position,
    background: options.background,
    color: options.color,
    lvlbar: options.lvlbar,
    lvlbarBg: options.lvlbarBg,
    member: message.guild.members.cache.get(userID)?.user
  })

  async function rankCard(message, options = []) {
    try {
      const {Canvas, loadImage, createCanvas} = require('canvas');

      function shortener(count) {
        const COUNT_ABBRS = [
          '',
          'k',
          'M',
          'B',
          'T',
          'Q',
          'Q+',
          'S',
          'S+',
          'O',
          'N',
          'D',
          'U'
        ]

        const i =
          0 === count ? count : Math.floor(Math.log(count) / Math.log(1000))
        let result = parseFloat((count / Math.pow(1000, i)).toFixed(2))
        result += `${COUNT_ABBRS[i]}`
        return result
      }

      const member = options.member

      const canvas = createCanvas(1024, 300),
      ctx = canvas.getContext('2d')

      const name = member.tag
      const noSymbols = (string) => string.replace(/[\u007f-\uffff]/g, '')

      let fsiz = '45px'
      if (message.guild.name.length >= 23) {
        fsiz = '38px'
      }
      if (message.guild.name.length >= 40) {
        fsiz = '28px'
      }
      if (message.guild.name.length >= 63) {
        fsiz = '22px'
      }

      let BackgroundRadius = '20',
        BackGroundImg =
          options.background ||
          'https://i.ibb.co/QQvMqf7/gradient.jpg',
        AttachmentName = 'rank.png',
        AttachmentDesc = 'Rank Card',
        Username = noSymbols(name),
        DrawLayerColor = '#000000',
        DrawLayerOpacity = 0.4,
        BoxColor = options.color || '#096DD1',
        LevelBarFill = options.lvlbar || '#ffffff',
        LevelBarBackground = options.lvlbarBg || '#ffffff',
        Rank = options.rank,
        LvlText = `УР.`,
        RankText = `РАНГ`,
        Lvl = `${shortener(options.level)}`,
        BarRadius = '20',
        TextXpNeded = '{current}/{needed}',
        CurrentXP = options.currentXP,
        NeededXP = options.neededXP

      ctx.beginPath()
      ctx.moveTo(0 + Number(BackgroundRadius), 0)
      ctx.lineTo(0 + 1080 - Number(BackgroundRadius), 0)
      ctx.quadraticCurveTo(0 + 1080, 0, 0 + 1080, 0 + Number(BackgroundRadius))
      ctx.lineTo(0 + 1080, 0 + 400 - Number(BackgroundRadius))
      ctx.quadraticCurveTo(
        0 + 1080,
        0 + 400,
        0 + 1080 - Number(BackgroundRadius),
        0 + 400
      )

      ctx.lineTo(0 + Number(BackgroundRadius), 0 + 400)
      ctx.quadraticCurveTo(0, 0 + 400, 0, 0 + 400 - Number(BackgroundRadius))
      ctx.lineTo(0, 0 + Number(BackgroundRadius))
      ctx.quadraticCurveTo(0, 0, 0 + Number(BackgroundRadius), 0)
      ctx.closePath()
      ctx.clip()
      ctx.fillStyle = '#000000'
      ctx.fillRect(0, 0, 1080, 400)
      let background = await loadImage(BackGroundImg)
      ctx.globalAlpha = 0.7
      ctx.drawImage(background, 0, 0, 1080, 400)
      ctx.restore()

      ctx.fillStyle = DrawLayerColor
      ctx.globalAlpha = DrawLayerOpacity
      ctx.globalAlpha = 1

      function RoundedBox(ctx, x, y, width, height, radius) {
        ctx.beginPath()
        ctx.moveTo(x + radius, y)
        ctx.lineTo(x + width - radius, y)
        ctx.quadraticCurveTo(x + width, y, x + width, y + radius)
        ctx.lineTo(x + width, y + height - radius)
        ctx.quadraticCurveTo(
          x + width,
          y + height,
          x + width - radius,
          y + height
        )
        ctx.lineTo(x + radius, y + height)
        ctx.quadraticCurveTo(x, y + height, x, y + height - radius)
        ctx.lineTo(x, y + radius)
        ctx.quadraticCurveTo(x, y, x + radius, y)
        ctx.closePath()
      }

      const avatar = await loadImage(member.displayAvatarURL({size: 256, extension: 'png' }))
      ctx.save();
      ctx.beginPath();
      ctx.arc(canvas.width/2 - canvas.width/4 - canvas.width/8 -  canvas.width/16 - 32 + 256/2, canvas.height/4 - 60 + 256/2, 256/2, 0, Math.PI * 2, true);
      ctx.closePath();
      ctx.clip();
  
      ctx.drawImage(avatar,  canvas.width/2 - canvas.width/4 - canvas.width/8 -  canvas.width/16 - 32, canvas.height/4 - 60, 256, 256);
  
      ctx.beginPath();
      ctx.arc(canvas.width/2 - canvas.width/4 - canvas.width/8 -  canvas.width/16 - 32, canvas.height/4 - 60, 256/2, 0, Math.PI * 2, true);
      ctx.clip();
      ctx.closePath();
      ctx.restore();

      ctx.save()
      ctx.textAlign = 'left'
      ctx.fillStyle = '#ffffff'
      ctx.shadowColor = '#000000'
      ctx.font = 'blod 70px Sans-Serif'
      ctx.fillText(Username, 320, 130)
      ctx.restore()

      ctx.save()
      RoundedBox(ctx, 320, 145, 660, 50, Number(BarRadius))
      ctx.strokeStyle = '#BFC85A22'
      ctx.stroke()
      ctx.clip()
      ctx.fillStyle = LevelBarBackground
      ctx.globalAlpha = 0.2
      ctx.fillRect(320, 145, 660, 50, 50)
      ctx.restore()

      const percent = (100 * CurrentXP) / NeededXP
      const progress = (percent * 660) / 100

      ctx.save()
      RoundedBox(ctx, 320, 145, progress, 50, Number(BarRadius))
      ctx.strokeStyle = '#BFC85A22'
      ctx.stroke()
      ctx.clip()
      ctx.fillStyle = LevelBarFill
      ctx.globalAlpha = 0.5
      ctx.fillRect(320, 145, progress, 50, 50)
      ctx.restore()

      ctx.save()
      ctx.textAlign = 'left'
      ctx.fillStyle = '#ffffff'
      ctx.globalAlpha = 0.8
      ctx.font = '30px Sans-Serif'
      ctx.fillText(LvlText, 320, 240)
      ctx.font = 'bold 50px Sans-Serif'
      ctx.fillText(Lvl, 355, 240)
      ctx.restore()

      ctx.save()
      ctx.textAlign = 'right'
      ctx.fillStyle = '#ffffff'
      ctx.shadowColor = '#000000'
      ctx.font = '30px Sans-Serif'
      ctx.fillText(RankText, 900, 240)
      ctx.font = 'bold 55px Sans-Serif'
      ctx.fillText('#' + Rank, 980, 240)
      ctx.restore()

      const latestXP = Number(CurrentXP) - Number(NeededXP)
      const textXPEdited = TextXpNeded.replace(/{needed}/g, shortener(NeededXP))
        .replace(/{current}/g, shortener(CurrentXP))
        .replace(/{latest}/g, latestXP)
      ctx.textAlign = 'center'
      ctx.fillStyle = '#474747'
      ctx.globalAlpha = 1
      ctx.font = '30px Sans-Serif'
      ctx.fillText(textXPEdited, 412, 180)

      const attachment = {
        attachment: canvas.toBuffer(),
        description: AttachmentDesc,
        name: AttachmentName
      }
      return attachment
    } catch (err) {
      console.log(`[XP] Error Occured. | rankCard | Error: ${err.stack}`)
    }
  }
}

module.exports = rank
